### PR TITLE
Refactor Project, Add Tests, Namespacing!

### PR DESCRIPTION
**Pull request by dcvz**
Hey guys,

As a big user of Sentry, I was patiently waiting for the new SDK to be out and now it is! 🎉 However, in my project I had some issues. Mostly around namespacing, I went in to fix the issues and in there thought some things could be improved:

- There's now better namespacing! Almost all classes/structs will fall under the `Sentry` namespace.  It works really well with Swift though somethings couldn't be moved because of the required interoperability with objc.
- I've now explicitly set access-control. There was many things that were public that probably shouldn't have been. There was also a divergence in code style (probably because of different people working on it).
- 47c463ba6b10df46da752eeac7ec41a7bd79e483 broke compatibility with `Breadcrumbs` in objc; however, it wasn't caught because objc test weren't testing for it. Now they are and the objc example also has them :) That same commit also raised other issues which I've addressed as in-line comments in this PR.

It's a big PR so let me know if there's any issues :)

### CODE REVIEW

File: Sources/Breadcrumb.swift [hunk 1]
```diff
@@ -8,81 +8,66 @@
...
-			.set("url", value: url)
-			.set("method", value: method)
-			.set("status_code", value: statusCode)
-			.set("reason", value: "reason")
-		
-		self.init(category: category, timestamp: timestamp, message: message, type: "http", level: level, data: httpData)
+		public var type: String?
+		public var message: String?
+		public let data: [String: AnyObject]
+		public let level: SentrySeverity // can't be optional because @objc can't handle optional enums
```
Comment (dcvz) [inline Sources/Breadcrumb.swift]: If we want interoperability with `objc` then we can't declare `SentrySeverity?` as such I've defaulted it to `.Info` which I think should be up to you guys to decide if that's OK. We could also introduce a new value to the enum like `.None` that we can treat as `nil`

File: Sources/Breadcrumb.swift [hunk 2]
```diff
@@ -8,81 +8,66 @@
...
+        public convenience init(category: String, timestamp: NSDate = NSDate(), message: String? = nil, level: SentrySeverity = .Info, data: [String: AnyObject]? = nil, to: String, from: String? = nil) {
+			let navigationData: [String: AnyObject] = (data ?? [:])
+				.set("to", value: to)
+				.set("from", value: from)
+
+			self.init(category: category, timestamp: timestamp, message: message, type: "navigation", level: level, data: navigationData)
+		}
+
+		/// Conveneince init for an "http" type breadcrumb (-999 workaround because @objc can't handle optional Int)
+        public convenience init(category: String, timestamp: NSDate = NSDate(), message: String? = nil, level: SentrySeverity = .Info, data: [String: AnyObject]? = nil, url: String, method: String, statusCode: Int = -999, reason: String? = nil) {
```
Comment (dcvz) [inline Sources/Breadcrumb.swift]: If we want this init to be available in `objc` then we can't have `Int?` along with the already mentioned `SentrySeverity?` as a workaround I've set the default to -999 then later check for it and treat it as `nil`

File: Sources/Breadcrumb.swift (hunk 1 above)
Comment (dcvz) [inline Sources/Breadcrumb.swift]: Looking at the [Interfaces documentation](https://docs.getsentry.com/hosted/clientdev/interfaces/) it already defaults to `.Info`on the server side so it should be ok :)

**Comment by ahti** [comment]
I don't know yet wether I think namespacing is necessary, but regardless, I think it would be much easier to review and discuss these changes if they were broken down into one pull-request per change, basically one for every bullet point in your list.

**Comment by dcvz** [comment]
I think namespacing should always happen in libraries because you don't want to cause conflicts with any types the user has in his project (i.e) I already had a `User` object. The commit about adding `Exceptions` got in by mistake but the example projects should be updated to reflect the changes in source so I think that one should be included in this PR.

File: Sources/Breadcrumb.swift [hunk 3]
```diff
@@ -8,81 +8,66 @@
...
-			.set("url", value: url)
-			.set("method", value: method)
-			.set("status_code", value: statusCode)
-			.set("reason", value: "reason")
-		
-		self.init(category: category, timestamp: timestamp, message: message, type: "http", level: level, data: httpData)
+		public var type: String?
+		public var message: String?
+		public var data: [String: AnyObject]
+		public var level: SentrySeverity // can't be optional because @objc can't handle optional enums
```
Comment (dcvz) [inline Sources/Breadcrumb.swift]: If we want interoperability with objc then we can't declare SentrySeverity? as such I've defaulted it to .Info which I think should be up to you guys to decide if that's OK. We could also introduce a new value to the enum like .None that we can treat as nil

File: Sources/Breadcrumb.swift (hunk 2 above)
Comment (dcvz) [inline Sources/Breadcrumb.swift]: Looking at the [Interfaces documentation](https://docs.getsentry.com/hosted/clientdev/interfaces/) it already defaults to `.Info`on the server side so it should be ok :)

File: Sources/Breadcrumb.swift (hunk 3 above)
Comment (dcvz) [inline Sources/Breadcrumb.swift]: If we want interoperability with objc then we can't declare SentrySeverity? as such I've defaulted it to .Info which I think should be up to you guys to decide if that's OK. We could also introduce a new value to the enum like .None that we can treat as nil

Comment (dcvz) [inline Sources/Breadcrumb.swift]: Looking at the [Interfaces documentation](https://docs.getsentry.com/hosted/clientdev/interfaces/) it already defaults to `.Info`on the server side so it should be ok :)

**Comment by joshdholtz** [comment]
@dcvz Thanks for the all the work here 🙌   I have two thoughts here...

1) I agree the the namespacing isn't necessary. Swift already has namespacing built in with the module so the module name can be used as a name space for any conflicting class names. So if you could remove that that would be great 😊 
2) Once namespacing is removed, can you change back `Sentry` to `SentryClient`. A majority (if not all) of the Sentry clients for other platforms use "client" in the class/struct/namespace and I would prefer to keep consistency with the rest of the clients.

Once those are changed I will take a look at the PR again 👌

**Comment by dcvz** [comment]
@joshdholtz I've made the changes and it's ready to review once more :)

File: Sources/Breadcrumb.swift [hunk 4]
```diff
@@ -8,81 +8,63 @@
...
-	
+
 	public var type: String?
 	public var message: String?
 	public var data: [String: AnyObject]
-	public var level: SentrySeverity?
-	
-	/// Creates a user
-	public init(category: String, timestamp: NSDate = NSDate(), message: String? = nil, type: String? = nil, level: SentrySeverity? = nil, data: [String: AnyObject]? = nil) {
+	public var level: SentrySeverity // can't be optional because @objc can't handle optional enums
```
Comment (joshdholtz) [inline Sources/Breadcrumb.swift]: Great use of comments 🙌

File: Sources/Breadcrumb.swift [hunk 5]
```diff
@@ -8,81 +8,63 @@
...
-	public convenience init(category: String, timestamp: NSDate = NSDate(), message: String? = nil, level: SentrySeverity? = nil, data: [String: AnyObject]? = nil, url: String, method: String, statusCode: Int? = nil, reason: String? = nil) {
-		
-		let httpData = (data ?? [:])
-			.set("url", value: url)
-			.set("method", value: method)
-			.set("status_code", value: statusCode)
-			.set("reason", value: "reason")
-		
+
+	/// Conveneince init for an "http" type breadcrumb (-999 workaround because @objc can't handle optional Int)
```
Comment (joshdholtz) [inline Sources/Breadcrumb.swift]: @dcvz Can it handle an optional `NSNumber`? Not sure better or worse if it can 😬

File: Sources/DSN.swift [hunk 6]
```diff
@@ -8,88 +8,80 @@
...
-		let value = headerParts.reduce([], combine: { (combined, keyValue) -> [String] in
-			guard let value = keyValue.1 else {
-				return combined
-			}
-			return combined + ["\(keyValue.0)=\(value)"]
-		}).joinWithSeparator(",")
-		
+
+		var ret: [String] = []
+		headerParts.filter() { $0.1 != nil }.forEach() { ret.append("\($0.0)=\($0.1!)") }
```
Comment (joshdholtz) [inline Sources/DSN.swift]: @dcvz You should be able to `.flatMap({$0.1})` instead of `.filter() { $0.1 != nil }`

Comment (joshdholtz) [inline Sources/DSN.swift]: Maybe something like...
```swift
var value = headerParts.flatMap({$0.1}).map({"\($0.0)=\($0.1!)"})
```

Didn't test but I like maps 😇

File: Sources/EventProperties.swift [hunk 7]
```diff
@@ -8,35 +8,26 @@
...
-        tags?.mergeValues(from: eventProperties.tags ?? [:])
-        extra?.mergeValues(from: eventProperties.extra ?? [:])
-        user = user ?? eventProperties.user
-	}
+
+    /*
+    Merges another EventProperties into this. Will only replace non-existing keys.
+    - Parameter from: EventProperties to merge
+    */
+    internal mutating func mergeProperties(from other: EventProperties) {
```
Comment (joshdholtz) [inline Sources/EventProperties.swift]: Mutating scares me but this seems legit 🤔

File: Sources/KSCrashHandler.swift [hunk 8]
```diff
@@ -6,128 +6,135 @@
...
 				}
-				
-				// Report crash time
+
+				// Extract crash timestamp
 				let timestamp: NSDate = {
 					var date: NSDate?
 					if let timestampStr = report["report"]?["timestamp"] as? String {
-						let dateFormatter = NSDateFormatter()
+						let dateFormatter: NSDateFormatter = NSDateFormatter()
```
Comment (joshdholtz) [inline Sources/KSCrashHandler.swift]: @dcvz Why the explicit type here?

File: Sources/KSCrashHandler.swift [hunk 9]
```diff
@@ -6,128 +6,135 @@
...
-				 * This is probably only temporary so that we can have some unique
-				 * info for the crash until we can dysymolicate logs on Sentry's API.
-				 */
-				let exceptionName = report["crash"]?["error"]??["mach"]??["exception_name"] as? String ?? "Fatal error"
-				
-				let backtrace = report["crash"]?["threads"] as? [CrashDictionary]
-				let crashedThread = backtrace?.filter { thread in
+                 * This is probably only temporary so that we can have some unique
+                 * info for the crash until we can desymbolicate logs on Sentry's API.
+                 */
```
Comment (joshdholtz) [inline Sources/KSCrashHandler.swift]: Tabbing here seems messed up a little?

File: Sources/KSCrashHandler.swift [hunk 10]
```diff
@@ -6,128 +6,135 @@
...
 					}
 					return exceptionName
 				}()
-				
+
 				/*
-				 * Generate event to sent up to API
-				 */
+                 * Generate event to sent up to API
+                 */
```
Comment (joshdholtz) [inline Sources/KSCrashHandler.swift]: Tabbing also messed up here

File: Sources/KSCrashHandler.swift [hunk 11]
```diff
@@ -137,15 +144,19 @@ extension SentryClient {
...
 		installation?.install()
 	}
-	
+
+
+	// MARK: - Private Helpers
+
 	private func updateUserInfo() {
-		var userInfo = CrashDictionary()
+		var userInfo: CrashDictionary = CrashDictionary()
```
Comment (joshdholtz) [inline Sources/KSCrashHandler.swift]: Can we remove this explicit type?

File: Sources/Request.swift [hunk 12]
```diff
@@ -10,47 +10,48 @@ import Foundation
...
-
-		// Logging cause yeah
 		if let body = NSString(data: data, encoding: NSUTF8StringEncoding) {
 			SentryLog.Debug.log("body = \(body)")
 		}
 		
-		// Creating request
-		let request = NSMutableURLRequest(URL: dsn.serverURL)
+		// Creating the request
+		let request: NSMutableURLRequest = NSMutableURLRequest(URL: dsn.serverURL)
```
Comment (joshdholtz) [inline Sources/Request.swift]: Explicit typ not needed

File: Sources/Request.swift [hunk 13]
```diff
@@ -10,47 +10,48 @@ import Foundation
...
 		request.HTTPMethod = "POST"
 		request.HTTPBody = data
 		
-		// Setting headers
-		let sentryHeader = dsn.xSentryAuthHeader
+		// Setting the headers
+		let sentryHeader: XSentryAuthHeader = dsn.xSentryAuthHeader
```
Comment (joshdholtz) [inline Sources/Request.swift]: Explicit typ not needed

File: Sources/Request.swift [hunk 14]
```diff
@@ -10,47 +10,48 @@ import Foundation
...
 		request.setValue(sentryHeader.value, forHTTPHeaderField: sentryHeader.key)
 		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
 		
 		// Creating data task
-		let config = NSURLSessionConfiguration.defaultSessionConfiguration()
-		let session = NSURLSession(configuration: config)
+		let config: NSURLSessionConfiguration = NSURLSessionConfiguration.defaultSessionConfiguration()
+		let session: NSURLSession = NSURLSession(configuration: config)
```
Comment (joshdholtz) [inline Sources/Request.swift]: Explicit types not needed

File: Sources/Sentry.swift [hunk 15]
```diff
@@ -68,67 +74,56 @@ public struct SentryInfo {
...
+	/// Creates a Sentry object iff a valid DSN is provided
 	@objc public convenience init?(dsnString: String) {
 		guard let dsn = DSN(dsnString) else {
 			return nil
 		}
 		self.init(dsn: dsn)
 	}
 	
-	/// The operating system this client is running on
-	public var os: OS {
```
Comment (joshdholtz) [inline Sources/Sentry.swift]: Was this not getting used anywhere??? 😱

File: Sources/DSN.swift (hunk 6 above)
Comment (dcvz) [inline Sources/DSN.swift]: I felt like using the `filter` was more verbose about the process, but I also prefer the `flatMap` in my day to day dev. I can change it if that's ok with you.

File: Sources/Sentry.swift [hunk 16]
```diff
@@ -68,67 +74,56 @@ public struct SentryInfo {
...
-				event.breadcrumbsSerialized = breadcrumbs.serialized
-				breadcrumbs.clear()
-			default: ()
-			}
+	/*
+	Reports given event to Sentry
+	- Parameter event: An event struct
+	- Parameter useClientProperties: Should the client's user, tags and extras also be reported (default is `true`)
+	*/
+	private func captureEvent(var event: Event, useClientProperties: Bool = true) {
```
Comment (joshdholtz) [inline Sources/Sentry.swift]: @dcvz This `var` needs to get 💣  It is getting depreciated in Swift 3

File: Sources/KSCrashHandler.swift (hunk 8 above)
Comment (dcvz) [inline Sources/KSCrashHandler.swift]: @joshdholtz habit. It's not needed in Swift but it speeds up compiling to specify types, like declaring `internal` can remove if it bothers you.

File: Sources/Sentry.swift (hunk 15 above)
Comment (dcvz) [inline Sources/Sentry.swift]: nope!

File: Sources/Sentry.swift (hunk 16 above)
Comment (dcvz) [inline Sources/Sentry.swift]: will change :)

File: Sources/Sentry.swift (hunk 15 above)
Comment (joshdholtz) [inline Sources/Sentry.swift]: Then good deletion 😉

File: Sources/KSCrashHandler.swift (hunk 8 above)
Comment (joshdholtz) [inline Sources/KSCrashHandler.swift]: It does a bit 😇  Not much bothers me but for some reason this one always gets to me haha

File: Sources/DSN.swift (hunk 6 above)
Comment (joshdholtz) [inline Sources/DSN.swift]: `flatMap` looks a bit cleaner to me 😇  I'm cool with either... I just fancy `flatMap` ❤️

**Comment by joshdholtz** [comment]
@dcvz All done with comments! Back to you 👼

File: Sources/DSN.swift (hunk 6 above)
Comment (dcvz) [inline Sources/DSN.swift]: actually I remember why I didn't use `flatMap` it doesn't work with the way we'd want it to when `flatMap`ing over [(String, String?)] 😞

**Comment by dcvz** [comment]
All done :)

**Comment by joshdholtz** [comment]
@dcvz Awesome! I will pull this down now and play around with it before merging... The code looks good though 👍

**Comment by dcvz** [comment]
Great! I'll make the PR that adds the Exception interface after :)

**Comment by joshdholtz** [comment]
@dcvz Everything seems to checkout so far! There weren't any "breaking changes" in this, correct? (asking for versioning reasons). I didn't see any changes to the `ViewController.swift` so I assume that this was mostly internal stuff and fixes for breadcrumbs, right?

**Comment by dcvz** [comment]
Yeah, there's no breaking changes :) It was mostly a refactor, I tried to keep functionality :)

**Comment by joshdholtz** [comment]
@dcvz Perfect! I will merge this in later today then... Thanks for help ❤️

**Comment by dcvz** [comment]
No problem :) big user of sentry so it's beneficial to us all. I'll pull up the other PR as soon as this is merged then @ahti can continue his work!

**Comment by dcvz** [comment]
any updates on this @joshdholtz ?

**Pull request merged**